### Task 1: DevStack Data Science Student Internship Program(DSIP)
### Technology used: Python
### IDE: Jupyter notebook
### Task objectives: Make a fake news detecting system from given data set using machine learning

### Importing libraries & othe stuff

### If you don't have numpy or pandas etc. Install it by using !pip install name --upgrade --quiet

In [1]:
import pandas as pd
import numpy as np
import itertools

#### Download the dataset with the given link in the pdf and place it in your cwd(current working directory)

In [2]:
news_df = pd.read_csv('news.csv')

In [3]:
news_df

Unnamed: 0                                              title  \
0           8476                       You Can Smell Hillary’s Fear   
1          10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2           3608        Kerry to go to Paris in gesture of sympathy   
3          10142  Bernie supporters on Twitter erupt in anger ag...   
4            875   The Battle of New York: Why This Primary Matters   
...          ...                                                ...   
6330        4490  State Department says it can't find emails fro...   
6331        8062  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332        8622  Anti-Trump Protesters Are Tools of the Oligarc...   
6333        4021  In Ethiopia, Obama seeks progress on peace, se...   
6334        4330  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2     U.S. Secretary of State John F. Kerry said Mon...  REAL  
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4     It's primary day in New York and front-runners...  REAL  
...                                                 ...   ...  
6330  The State Department told the Republican Natio...  REAL  
6331  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE  
6332   Anti-Trump Protesters Are Tools of the Oligar...  FAKE  
6333  ADDIS ABABA, Ethiopia —President Obama convene...  REAL  
6334  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL  

[6335 rows x 4 columns]

#### As you can see that the given dataset contains 6335 rows and 4 columns. **The 4th column is the important one**. 

#### Viewing whole dataset
#### (For that we will use Ipython.display)

In [4]:
# Ipython library is useful here set max rows as 10k because given rows in dataset is <7k
from IPython.display import display
with pd.option_context('display.max_rows', 10000):
    display(news_df)

Unnamed: 0                                              title  \
0           8476                       You Can Smell Hillary’s Fear   
1          10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2           3608        Kerry to go to Paris in gesture of sympathy   
3          10142  Bernie supporters on Twitter erupt in anger ag...   
4            875   The Battle of New York: Why This Primary Matters   
5           6903                                        Tehran, USA   
6           7341  Girl Horrified At What She Watches Boyfriend D...   
7             95                  ‘Britain’s Schindler’ Dies at 106   
8           4869  Fact check: Trump and Clinton at the 'commande...   
9           2909  Iran reportedly makes new push for uranium con...   
10          1357  With all three Clintons in Iowa, a glimpse at ...   
11           988  Donald Trump’s Shockingly Weak Delegate Game S...   
12          7041  Strong Solar Storm, Tech Risks Today | S0 News...   
13          7623       10 Ways America Is Preparing for World War 3   
14          1571                   Trump takes on Cruz, but lightly   
15          4739                         How women lead differently   
16          7737  Shocking! Michele Obama & Hillary Caught Glamo...   
17          8716  Hillary Clinton in HUGE Trouble After America ...   
18          3304  What's in that Iran bill that Obama doesn't like?   
19          3078  The 1 chart that explains everything you need ...   
20          2517  The slippery slope to Trump’s proposed ban on ...   
21         10348  Episode #160 – SUNDAY WIRE: ‘Hail to the Deplo...   
22           778  Hillary Clinton Makes A Bipartisan Appeal on S...   
23          3300  New Senate majority leader’s main goal for GOP...   
24          6155              ‘Inferno’ and the Overpopulation Myth   
25           636  Anti-Trump forces seek last-ditch delegate revolt   
26           755  Sanders Trounces Clinton in W. Va. -- But Will...   
27           626  Donald Trump Is Changing His Campaign Slogan t...   
28           691  Pure chaos: Donald Trump’s campaign management...   
29          5743  Syrian War Report – November 1, 2016: Syrian M...   
30          1787                     GOP insiders: Carly crushed it   
31          7808  Jeffrey Sewell et al. : Metabiology face to fa...   
32          6484         Why It’s Necessary To Relax Into A Stretch   
33          7385  Brexit Encourages UK to Trade With Non-EU Stat...   
34          9016  2016 interview with Socrates: Another day of L...   
35          9324  Mike Pence Drapes Shawl Over Immodest Lady Jus...   
36          8983       First Ever Hindu Woman Elected into Congress   
37          8965  Donald Groped Hillary in 2005! Trump and Weine...   
38          5580  Ex-Assistant FBI Director: Clintons Are a Crim...   
39          9757  Hillary Wants Aggressively Interventionist For...   
40          1967  Both parties want to craft populist messages f...   
41           431  First Take: Wall Street bids goodbye to June hike   
42          5955  Real Disclosure! Secret Alien Base Found In Mo...   
43          7455  Homeless Woman Protects Trump’s Walk of Fame S...   
44          5224  With 3:20 a.m. tweet storm Saturday, Clinton c...   
45          7793  220 ‘Significant’ Pipeline Spills Already This...   
46          2777  Obama makes the right call to tough it out in ...   
47           587  Senate race rankings: Dems attack as GOP lays ...   
48          9403  ‘He didn’t know the boy didn’t want to be rape...   
49          5931  Pieczenik “Rogue FBI Agents and Wikileaks are ...   
50          3027    American politics has reached peak polarization   
51          4377  Anti-Muhammad cartoon contest: Free speech or ...   
52          5937  3 Effects of Substance Abuse on Individual, Fa...   
53          7277  Tree Shaped Vertical Farms That Grow 24 Acres ...   
54          5521               New Comment Features have been Added   
55          3627         World

In [5]:
news_df.shape

(6335, 4)

#### Dropping useless columns 

In [7]:
news_df.drop(['Unnamed: 0'],axis=1, inplace=True)

In [8]:
news_df.head(5)

title  \
0                       You Can Smell Hillary’s Fear   
1  Watch The Exact Moment Paul Ryan Committed Pol...   
2        Kerry to go to Paris in gesture of sympathy   
3  Bernie supporters on Twitter erupt in anger ag...   
4   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

#### Checking for null/ NaN or invalid data in the dataset

In [9]:
# checking for null values
news_df.isnull().sum()

title    0
text     0
label    0
dtype: int64

In [10]:
# checking for NaN values
news_df.isna().sum()

title    0
text     0
label    0
dtype: int64

#### We have no inconsistent values hence we can proceed

In [11]:
labels = news_df.label

In [12]:
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

#### do this if u don't have sklearn 
#### !pip install scikit-learn

#### We are going to use sklearn here if you want more information of sklearn.model_selection use this ['https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html']

In [13]:
from sklearn.model_selection import train_test_split

#### Training

### The need for training data you may refer 'https://appen.com/blog/training-data/'


In [14]:
# basic concepts of ML taking test size 0.25 and random state 20
x_train,x_test,y_train,y_test = train_test_split(news_df["text"],labels,test_size=0.25,random_state = 20)

In [15]:
x_train.head()

832     Photo by The U.S. Army | CC BY 2.0 \n\nHere is...
3983    An admitted Al Qaeda agent released this month...
5419    Canada's Liberal Party won a decisive majority...
4616    FBI Releases Files on Bill Clinton's Cash for ...
4289    There are just days left until Americans head ...
Name: text, dtype: object

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier

#### Intializing the vectorizer

In [17]:
vc = TfidfVectorizer(stop_words = 'english', max_df = 0.7)

#### Fit and Transform

In [18]:
tf_train  = vc.fit_transform(x_train)
tf_test = vc.transform(x_test)

#### Intializing passive aggressive classifier

In [19]:
pgc = PassiveAggressiveClassifier(max_iter=50)
pgc.fit(tf_train,y_train)

PassiveAggressiveClassifier(max_iter=50)

#### Predicition

#### For more info on prediction and passive aggresive classifier use 'https://www.geeksforgeeks.org/passive-aggressive-classifiers/'

In [20]:
from sklearn.metrics import accuracy_score, confusion_matrix
pred = pgc.predict(tf_test)

In [21]:
score = accuracy_score(y_test,pred)

In [22]:
print("Accuracy is {} %".format(round(score*100,2)))

Accuracy is 94.32 %


#### confusion matrix

#### A confusion matrix is a table that is often used to describe the performance of a classification model (or "classifier") on a set of test data for which the true values are known.

In [31]:
confusion_matrix(y_test,pred,labels=['FAKE','REAL'])

array([[776,  34],
       [ 52, 722]], dtype=int64)

#### Saving the file

#### You can save the file on your local system or save in kaggle Jovian Colab etc depending on whatever you are using